### DISCUSSION TITLE
SMARTS reaction produces duplicated identical products

### DISCUSSION THREAD

**Discussion by GLPG-GT** [rdkit/rdkit#6277] [Q&A] [answered]
I needed to make a reaction to 'clip' aromatic boronic acids and boronates.

This is what I came up with:

```
clip_boronate_2_SMARTS = '[c:0]-B(-O)-O>>[c:0]-[2*]'
clip_boronate_2 = AllChem.ReactionFromSmarts(clip_boronate_2_SMARTS)
clip_boronate_2
```
![download](https://user-images.githubusercontent.com/96192927/230576075-1e8a6873-3f18-4046-8df9-730eb4277d6d.png)

However, when I apply it to a reactant, I get twice the same molecule:

```
prod = clip_boronate_2.RunReactants((Chem.MolFromSmiles('OB(O)c1ccccc1'),))
prod_mols_list = []
for p in prod :
    prod_mols_list.extend(list(p))
Draw.MolsToGridImage(prod_mols_list, molsPerRow = 2)
```
![download](https://user-images.githubusercontent.com/96192927/230576253-e679c998-aee1-4e04-9817-92548e4e5aa2.png)

Does anybody know why that is?
When I clip a stannane, I get 6 times the same product:

```
clip_stannane_2_SMARTS = '[c:0]-[Sn](-[#6])(-[#6])-[#6]>>[c:0]-[2*]'
clip_stannane_2 = AllChem.ReactionFromSmarts(clip_stannane_2_SMARTS)
clip_stannane_2
prod = clip_stannane_2.RunReactants((Chem.MolFromSmiles('CC[Sn](CC)(CC)c1ccccc1'),))
prod_mols_list = []
for p in prod :
    prod_mols_list.extend(list(p))
Draw.MolsToGridImage(prod_mols_list, molsPerRow = 2)
```
![download](https://user-images.githubusercontent.com/96192927/230576450-2ad7aa4a-771c-4c49-9249-12f8e5e2c614.png)

Obviously I could fix this a posteriori by converting the product molecules to SMILES and removing duplicates:
```
prod_mols_list = []
for p in prod :
    prod_mols_list.append([Chem.MolToSmiles(pi) for pi in list(p)])
prod_mols_list

[['[2*]c1ccccc1'],
 ['[2*]c1ccccc1'],
 ['[2*]c1ccccc1'],
 ['[2*]c1ccccc1'],
 ['[2*]c1ccccc1'],
 ['[2*]c1ccccc1']]
```
but first I would like to understand if this behaviour is due to my incorrect SMARTS definition or anything else I am doing wrong, or instead something unexpected.

Any ideas?
Thanks

**Comment by bp-kelley** [comment, marked as answer]
Your pattern for boronic acid is symmetric and matches twice.

```
>>> sma = "[c:0]-B(-O)-O"
>>> pat = Chem.MolFromSmarts(sma)
>>> len(Chem.MolFromSmiles('OB(O)c1ccccc1').GetSubstructMatches(pat, uniquify=False))
2
```

So two products will be made.  You can resolve this by using recursive smarts that only match the boron in the boronic acid, but these can be a bit mind bending.

```
# boronic acid pattern that reports a match only on the Boron
#  [$(B(-!@[#6])(O)(O))]
clip_boronate_2_SMARTS = '[c:1][$(B(-!@[#6])(O)(O))]>>[c:1]-[2*]'
clip_boronate_2 = AllChem.ReactionFromSmarts(clip_boronate_2_SMARTS)

prod = clip_boronate_2.RunReactants((Chem.MolFromSmiles('OB(O)c1ccccc1'),))
prod_mols_list = []
for p in prod :
    for m in p:
        print(Chem.MolToSmiles(m))
```

For more details see section 4.4:

https://www.daylight.com/dayhtml/doc/theory/theory.smarts.html

We also have a list of functional group patterns available:

```
Halogen [$([F,Cl,Br,I]-&!@[#6])&!$([F,Cl,Br,I]-&!@C-&!@[F,Cl,Br,I])&!$([F,Cl,Br,I]-[C,S]=[O,S,N])]
Halogen.Aliphatic [$([F,Cl,Br,I]-&!@C)&!$([F,Cl,Br,I]-&!@C-&!@[F,Cl,Br,I])]
Halogen.Aromatic [F,Cl,Br,I;$(*-&!@c)]
Halogen.Bromine [Br&$(Br-&!@[#6])&!$(Br-&!@C-&!@[F,Cl,Br,I])&!$(Br-[C,S]=[O,S,N])]
Halogen.Bromine.Aliphatic [Br&$(Br-&!@C)&!$(Br-&!@C-&!@[F,Cl,Br,I])&!$(Br-[C,S]=[O,S,N])]
Halogen.Bromine.Aromatic [Br&$(Br-&!@c)]
Halogen.Bromine.BromoKetone [Br&$(Br-[C&H2]-C(=O)-[#6])]
Halogen.NotFluorine [$([Cl,Br,I]-&!@[#6])&!$([Cl,Br,I]-&!@C-&!@[F,Cl,Br,I])&!$([Cl,Br,I]-[C,S]=[O,S,N])]
Halogen.NotFluorine.Aliphatic [$([Cl,Br,I]-&!@C)&!$([Cl,Br,I]-&!@C-&!@[F,Cl,Br,I])&!$([Cl,Br,I]-[C,S]=[O,S,N])]
Halogen.NotFluorine.Aromatic [$([Cl,Br,I]-&!@c)]
Isocyanate [$(N-&!@[#6])]=&!@C=&!@O
Isocyanate.Aliphatic [$(N-&!@C)]=&!@C=&!@O
Isocyanate.Aromatic [$(N-&!@c)]=&!@C=&!@O
Nitro [N&H0&$(N-[#6])&D3](=[O&D1])~[O&D1]
Nitro.Aliphatic [N&H0&$(N-C)&D3](=[O&D1])~[O&D1]
Nitro.Aromatic [N&H0&$(N-c)&D3](=[O&D1])~[O&D1]
SulfonylChloride [$(S-&!@[#6])](=O)(=O)Cl
SulfonylChloride.Aliphatic [$(S-&!@C)](=O)(=O)Cl
SulfonylChloride.Aromatic [$(S-&!@c)](=O)(=O)Cl
TerminalAlkyne [C&$(C#[C&H1])]
```

**Reply by GLPG-GT** [reply]
Thanks!
Indeed, you're right, it's mind-bending.
E.g. I don't get why there have to be two carbon atoms in `clip_boronate_2_SMARTS = '[c:1][$(B(-!@[#6])(O)(O))]>>[c:1]-[2*]'`.
But OK, I will try to understand this recursive syntax and apply it to future cases, to avoid redundancy; or maybe just reduce it a posteriori.
BTW sometimes I do want symmetry, and it works well, e.g. when I want to break a biaryl carbon-carbon bond so that one moiety has the electrophile role (1*) and one the nucleophile role (2*):
`'[c:0]-[c:1]>>[c:0]-[1*].[c:1]-[2*]'`
then if I apply this to a biaryl, I get both possible disconnections.
I just did not see how my boronate was symmetric, when the bond I was breaking was a [c]-B bond, not a [c]-[c] one.
